src/test/java/io/netty/incubator/codec/http3/EmbeddedQuicChannel.java

```diff
@@ -0,0 +1,275 @@
+import io.netty.util.concurrent.Promise;
+
+import java.util.Collection;
+import java.util.EnumMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+import static java.util.Collections.unmodifiableCollection;
+
+final class EmbeddedQuicChannel extends EmbeddedChannel implements QuicChannel {
+    private static final AtomicLongFieldUpdater<EmbeddedQuicChannel> streamIdUpdater =
+            AtomicLongFieldUpdater.newUpdater(EmbeddedQuicChannel.class, "streamId");
+
+    private volatile long streamId;
+    private final Map<QuicStreamType, Long> peerAllowedStreams = new EnumMap<>(QuicStreamType.class);
+    private final AtomicBoolean closed = new AtomicBoolean();
+    private final ConcurrentLinkedQueue<Integer> closeErrorCodes = new ConcurrentLinkedQueue<>();
+    private QuicChannelConfig config;
+
+    EmbeddedQuicChannel(ChannelHandler... handlers) {
+        super(handlers);
+    }
+
+    EmbeddedQuicChannel(Channel parent, ChannelId channelId, boolean register, boolean hasDisconnect,
+                        ChannelHandler... handlers) {
+        super(parent, channelId, register, hasDisconnect, handlers);
+    }
+
+    @Override
+    public QuicChannelConfig config() {
+        if (config == null) {
+            config = new EmbeddedQuicChannelConfig(super.config());
+        }
+        return config;
+    }
+
+    @Override
+    public QuicChannel flush() {
+        super.flush();
+        return this;
+    }
+
+    @Override
+    public QuicChannel read() {
+        super.read();
+        return this;
+    }
+
+    @Override
+    public long peerAllowedStreams(QuicStreamType type) {
+        return peerAllowedStreams.getOrDefault(type, Long.MAX_VALUE);
+    }
+
+    public void peerAllowedStreams(QuicStreamType type, long peerAllowedStreams) {
+        this.peerAllowedStreams.put(type, peerAllowedStreams);
+    }
+
+    @Override
+    public Future<QuicStreamChannel> createStream(QuicStreamType type, ChannelHandler handler,
+                                                  Promise<QuicStreamChannel> promise) {
+        promise.setSuccess(new EmbeddedQuicStreamChannel(this, true, type,
+                streamIdUpdater.getAndAdd(this, 2), handler));
+        return promise;
+    }
+
+    @Override
+    public ChannelFuture close(boolean applicationClose, int error, ByteBuf reason, ChannelPromise promise) {
+        closeErrorCodes.add(error);
+        if (closed.compareAndSet(false, true)) {
+            promise.addListener(__ -> reason.release());
+        } else {
+            reason.release();
+        }
+        return close(promise);
+    }
+
+    @Override
+    public Future<QuicConnectionStats> collectStats(Promise<QuicConnectionStats> promise) {
+        promise.setFailure(new UnsupportedOperationException("Collect stats not supported for embedded channel."));
+        return promise;
```

Comment:
nit: you could merge the above lines as `promise.setSuccess` returns the promise as well